android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt

```diff
@@ -595,6 +598,59 @@ class OkHttpTest {
     }
   }
 
+  @Test
+  fun testCachedRequest() {
+    assumeNetwork()
+
+    val ctxt = InstrumentationRegistry.getInstrumentation().targetContext.applicationContext
+
+    val cacheSize = 1L * 1024 * 1024 // 1MB
+    val cache = Cache(ctxt.cacheDir.resolve("testCache"), cacheSize)
+
+    try {
+      client = client.newBuilder()
+          .cache(cache)
+          .build()
+
+      val request = Request.Builder()
+          .url("https://httpbin.org/cache/3")
+          .build()
+
+      client.newCall(request)
+          .execute()
+          .use {
+            assertEquals(200, it.code)
+            assertNull(it.cacheResponse)
+            assertNotNull(it.networkResponse)
+          }
+
+      client.newCall(request)
+          .execute()
+          .use {
+            assertEquals(200, it.code)
+            assertNotNull(it.cacheResponse)
+            assertNull(it.networkResponse)
+          }
+
+      Thread.sleep(5000)
+
+      client.newCall(request)
+          .execute()
+          .use {
+            assertEquals(200, it.code)
+            // TODO understand this?
```

Comment:
It means the network was used to check whether the cache was fresh

Reply:
I think this always surprises me, that Caches don't stop the network request got valid content. :)

I need to read up more on our heuristic caching.

https://www.mnot.net/blog/2009/02/24/unintended_caching
https://paulcalvano.com/index.php/2018/03/14/http-heuristic-caching-missing-cache-control-and-expires-headers-explained/

Reply:
> It means the network was used to check whether the cache was fresh

Actually I'm convinced it is the opposite case, the cache was fetched, found stale and then a network call was made.